omp/matrix/csr_kernels.cpp

```diff
@@ -363,6 +364,175 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    auto orig_row_ptrs = orig->get_const_row_ptrs();
+    auto orig_col_idxs = orig->get_const_col_idxs();
+    auto orig_vals = orig->get_const_values();
+    auto cp_row_ptrs = column_permuted->get_row_ptrs();
+    auto cp_col_idxs = column_permuted->get_col_idxs();
+    auto cp_vals = column_permuted->get_values();
+    auto num_nnz = orig->get_num_stored_elements();
+    size_type num_rows = orig->get_size()[0];
+    size_type num_cols = orig->get_size()[1];
+
+    std::vector<IndexType> pindx(&perm[0], &perm[0] + num_cols);
+    std::vector<size_type> ind(num_nnz);
+    std::vector<IndexType> inv_pindx(pindx);
+#pragma omp parallel for
+    for (size_type k = 0; k < pindx.size(); ++k) {
+        inv_pindx[pindx[k]] = k;
+    }
+#pragma omp parallel for
+    for (size_type row = 0; row < num_rows; ++row) {
+        cp_row_ptrs[row] = orig_row_ptrs[row];
+        for (size_type k = static_cast<size_type>(orig_row_ptrs[row]);
+             k < static_cast<size_type>(orig_row_ptrs[row + 1]); ++k) {
+            cp_col_idxs[k] = inv_pindx[orig_col_idxs[k]];
+            cp_vals[k] = orig_vals[k];
+        }
+    }
+    cp_row_ptrs[num_rows] = orig_row_ptrs[num_rows];
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_CSR_COLUMN_PERMUTE_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void inverse_row_permute(std::shared_ptr<const OmpExecutor> exec,
+                         const Array<IndexType> *permutation_indices,
+                         matrix::Csr<ValueType, IndexType> *row_permuted,
+                         const matrix::Csr<ValueType, IndexType> *orig)
+{
+    auto perm = permutation_indices->get_const_data();
+    auto orig_row_ptrs = orig->get_const_row_ptrs();
+    auto orig_col_idxs = orig->get_const_col_idxs();
+    auto orig_vals = orig->get_const_values();
+    auto rp_row_ptrs = row_permuted->get_row_ptrs();
+    auto rp_col_idxs = row_permuted->get_col_idxs();
+    auto rp_vals = row_permuted->get_values();
+    size_type num_rows = orig->get_size()[0];
+    size_type num_nnz = orig->get_num_stored_elements();
+
+    size_type cur_ptr = 0;
+    rp_row_ptrs[0] = cur_ptr;
+    std::vector<size_type> orig_num_nnz_per_row(num_rows, 0);
+    std::vector<IndexType> pindx(&perm[0], &perm[0] + num_rows);
+    std::vector<IndexType> inv_pindx(pindx);
+#pragma omp parallel for
+    for (size_type k = 0; k < pindx.size(); ++k) {
+        inv_pindx[pindx[k]] = k;
+    }
+#pragma omp parallel for
+    for (size_type row = 0; row < num_rows; ++row) {
+        orig_num_nnz_per_row[row] = orig_row_ptrs[row + 1] - orig_row_ptrs[row];
+    }
+    for (size_type row = 0; row < num_rows; ++row) {
+        rp_row_ptrs[row + 1] =
+            rp_row_ptrs[row] + orig_num_nnz_per_row[inv_pindx[row]];
+    }
+    rp_row_ptrs[num_rows] = orig_row_ptrs[num_rows];
+#pragma omp parallel for
+    for (size_type row = 0; row < num_rows; ++row) {
+        auto new_row = perm[row];
```

Comment:
You can remove this line or change it to 
```suggestion
        auto new_row = inv_pindx[row];
```
and use the value below